api/v1alpha4/openstackcluster_webhook.go

```diff
@@ -64,10 +68,34 @@ func (r *OpenStackCluster) ValidateCreate() error {
 func (r *OpenStackCluster) ValidateUpdate(old runtime.Object) error {
 	var allErrs field.ErrorList
 
+	newOpenStackCluster, err := runtime.DefaultUnstructuredConverter.ToUnstructured(r)
+	if err != nil {
+		return apierrors.NewInvalid(GroupVersion.WithKind("OpenStackCluster").GroupKind(), r.Name, field.ErrorList{
+			field.InternalError(nil, errors.Wrap(err, "failed to convert new OpenStackCluster to unstructured object")),
+		})
+	}
+	oldOpenStackCluster, err := runtime.DefaultUnstructuredConverter.ToUnstructured(old)
+	if err != nil {
+		return apierrors.NewInvalid(GroupVersion.WithKind("OpenStackCluster").GroupKind(), r.Name, field.ErrorList{
+			field.InternalError(nil, errors.Wrap(err, "failed to convert old OpenStackCluster to unstructured object")),
+		})
+	}
+
 	if r.Spec.IdentityRef != nil && r.Spec.IdentityRef.Kind != defaultIdentityRefKind {
 		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "identityRef", "kind"), "must be a Secret"))
 	}
 
+	newOpenStackClusterSpec := newOpenStackCluster["spec"].(map[string]interface{})
+	oldOpenStackClusterSpec := oldOpenStackCluster["spec"].(map[string]interface{})
+
+	// allow changes to ControlPlaneEndpoint
+	delete(oldOpenStackClusterSpec, "controlPlaneEndpoint")
+	delete(newOpenStackClusterSpec, "controlPlaneEndpoint")
```

Comment:
Looking at the cluster and machine controllers, we MUST be able to set controlPlaneEndpoint if it is not currently set, but I don't think we support changing it if it is already set.

This change would be an improvement over the current situation so I wouldn't say hold the change for this, but it might be good to include this nuance for completeness if you can.

Reply:
yeah.. the reason is we will patch the spec in our code in controller
without those 2 lines , the deployment of cluster will fail..

Reply:
Right. I was thinking something like:

```golang
  // Only allow controlPlaneEndpoint to be updated if it has not been set previously
  if testThatAllFieldsAreEmptySomehow(oldOpenStackClusterSpec["controlPlaneEndpoint"]) {
    delete(oldOpenStackClusterSpec, "controlPlaneEndpoint")
    delete(newOpenStackClusterSpec, "controlPlaneEndpoint")
  }
```

Reply:
ok, will take this a TODO (with other webhook like openstackmachine into one PR)

thanks~